### PR TITLE
feat: spellchecking API.

### LINKED ISSUE #724: Introduce a spellchecker in rust and implement lua bindings for it
## Description

To facilitate a blightmud spellchecker we'll need a spellchecking library. Hunspell is one option and it looks like creates exist for this already.

## Steps

- Add a spellchecking library
- Create a lua module and bindings for said spellchecker

### PR DESCRIPTION
**Pull request by cpu**
#### Description

This branch adds a new Lua module, `spellcheck`, with three functions:

* `spellcheck.init(aff_path, dict_path)` - initializes a spellchecker with the paths to a [Hunspell][0] compatible affix file and dictionary file. Must be called before any other spellcheck functions or an error will be thrown.
* `spellcheck.check(word) -> bool` - returns true if the word is in the dictionary that the spellchecker was initialized with. Errs if the spellchecker hasn't been initialized.
* `spellcheck.suggest(word) -> table` - if the given word is misspelled, returns a table array of suggested replacements in order of likelihood.

Presently these features are implemented with a backend that relies on the [`hunspell-rs` crate][1], which in turn uses [`hunspell-sys`][2] to interact with [libhunspell][0]. Hunspell is used extensively (e.g. by LibreOffice, Firefox) so compatible dictionaries/affix files are easy to source and the results are competitive with popular software. To avoid adding a new external dependency we use the "bundled" feature of `hunspell-sys` to have that crate bring along its own statically linked libhunspell. This means no changes are required for CI or the release process.

In the future the [`zspell`][3] crate looks like a promising pure Rust alternative to `hunspell-rs` that supports the Hunspell format and API. I elected not to use that crate for the short term because it doesn't yet have a stable API for providing suggestions and that's a core feature for a spellchecker. It should be possible to switch to this crate transparent to end users once it's feature complete.

For an example interactive spellchecker plugin built with this new Lua module, see [Blightspell][4].

Resolves #724.

[0]: http://hunspell.github.io/
[1]: https://crates.io/crates/hunspell-rs
[2]: https://crates.io/crates/hunspell-sys
[3]: https://crates.io/crates/zspell
[4]: https://github.com/cpu/blightspell

#### Demo

Here's a quick recorded demo of Blightspell in action:

[asciicast](https://asciinema.org/a/ZvMTNWBIOk1ib84isRUxrUHBu)

### CODE REVIEW

File: resources/help/spellcheck.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Spellcheck
+
+This module offers methods to perform spellchecking using a Hunspell 
+compatible dictionary. Before use, the spellcheck module must be
+initialized by providing paths to an AFF file, and a dictionary file.
+
```
Comment (LiquidityC) [inline resources/help/spellcheck.md:6]: A minor extension of this intro that makes it clear that this module is **not** a fully fledged spellchecker.
Feel free to promote your own plugin here :wink: 

Something like:

> This module is not a complete spellchecking solution. It provides methods for checking the spelling of a word and finding spelling suggestions for a particular word for use within scripts.
> For a more complete spellchecker there are plugins available. Such as:
>
> - Your solution with link

File: resources/help/spellcheck.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+- `dict_path`   path to a Hunspell dictionary file.
+
+##
+
+***spellcheck.check(word) -> bool***
+Checks whether the given word exists in the dictionary. If called
+before `spellcheck.init` an error will be produced.
+
+- `word`    A potentially misspelled word.
+
```
Comment (LiquidityC) [inline resources/help/spellcheck.md:25]: Add:
- Returns true if the spelling is correct otherwise false

File: resources/help/spellcheck.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+##
+
+***spellcheck.suggest(word) -> table***
+Returns a table of suggested replacements for a misspelled word. 
+If called before `spellcheck.init` an error will be produced.
+
+- `word`    A potentially misspelled word.
+- `table`   An array consisting of candidate replacement words, in
```
Comment (LiquidityC) [inline resources/help/spellcheck.md:33]: Change this last list entry into this format:
> Returns a bla bla bla...

File: src/lua/spellcheck.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    pub fn init(&mut self, aff_path: &str, dict_path: &str) {
+        self.hunspell
+            .replace(HunspellSafe::from(Hunspell::new(aff_path, dict_path)));
+    }
+
+    fn check_initialized(&self) -> Result<()> {
+        if self.hunspell.is_none() {
+            return Err(anyhow!("spellchecker not initialized"));
```
Comment (LiquidityC) [inline src/lua/spellcheck.rs:25]: Change to `Err(anyhow!("spellchecker not initialized"))`

You can put the `Ok(())` in an else block or use the `match` pattern.

**Review by LiquidityC — CHANGES_REQUESTED** [review]

**Comment by cpu** [comment]
@LiquidityC I integrated your feedback, added unit test coverage, and marked the PR ready for review. Let me know if I can do anything else.

**Review by LiquidityC — APPROVED** [review]

**Pull request merged**